### DISCUSSION TITLE
有毒的迴圈

### DISCUSSION THREAD

**Discussion by eason8479** [1101-GS4719-C/Python-Community-GS4719-C#808] [Show and tell]
## 前言
以下改編自個人經歷，純屬個人經驗分享
希望能在某天，當你遇到相似的問題時，能幫到你

解法肯定不是最佳解，歡迎各路大神挑戰

## 前情提要
故事發生在暑假的某天，我在為ITSA(一個台灣的程式考試)準備刷題目，在刷到 "[C_GM01-易] 計算三角形個數" 後就發生了下列的故事

## 題目
[ITSA 題庫連結](https://e-tutor.itsa.org.tw/e-Tutor/mod/programming/view.php?id=1060)  需申請帳號
```
問題描述 ：
根據三角不等式，三角形的任兩邊和要大於第三邊。假設現在有n根木棍，長度分別為1, 2, 3, …, n。你要挑出其中三根來拼成一個三角形，請問你可以拼出多少種長度不全相同的三角形？令此數目表示成T( n )。

輸入說明 ：
程式的輸入包含兩行數字，第一行包含一個正整數k，1 ≤ k ≤ 10，代表第二行有k個正整數n1, n2, ..., nk，1 ≤ ni ≤ 10000，而此k個正整數間以空格隔開。

輸出說明 ：
針對每一個測試資料ni，輸出一個答案Ti，共輸出k列答案。

範例輸入：
2 
5 8

範例輸出：
3 
22

```
## 第一次解題
看完問題後，我便開始不停的想，不停的試，最後終於完成了要求
思路也不算太複雜大致上就是
一個函式負責判斷三角形
```python
def triangle_test(a,b,c):  
    if (a+b >= c) and (a+c > b) and (b+c > a):  
        if (abs(a-b) < c) and (abs(a-c) < b) and (abs(b-c) < a):  
            return True     
    return False
```

另一個函式把n根木棍取3根的所有可能都跑一遍
```python
def count_triangle(n):  
    num = 0  
  
    for i in range (1,n-1):#  i 為第一根的長度
        for j in range(i+1,n):#  j 為第二根的長度
            for k in range (j+1,n+1): #  k 為第三根的長度
                if (triangle_test(i,j,k)):  
                    num += 1      
    print (num)
```

最後再加上一些小小的細節

```python
def triangle_test(a,b,c):  
    if (a+b >= c) and (a+c > b) and (b+c > a):  
        if (abs(a-b) < c) and (abs(a-c) < b) and (abs(b-c) < a):  
            return True     
    return False  
  
def count_triangle(n):  
    num = 0  
  
    for i in range (1,n-1):  
        for j in range(i+1,n):  
            for k in range (j+1,n+1):  
                if (triangle_test(i,j,k)):  
                    num += 1      
    print (num)  
  
turn = int(input())  
li = input().split()  
  
for i in range (turn):  
    count_triangle(int(li[i])) 
```
正當我一邊陶醉在自己打好的程式碼，一邊從一旁的馬克杯中啜飲著一口白開水
等待我的卻是3綠2紅(三題對兩題錯)的結果，為什麼呢?我也很好奇

## 第二次解題
經過多方的查找，我才知道ITSA是有時間限制的，運行時間大於10秒就會直接算你錯。
要是程式慢了該怎麼辦呢? 直上4090? Intel 13gen? Ryzen 7950?
很可惜我沒錢 (誤...等等好像也是
但課金是沒用的，因為用的是考試方的電腦，所以用戶這邊再快都沒用。

仔細思(ㄅㄥ)索(ㄎㄨㄟˋ)了一會兒後，我發現了問題的所在
for 迴圈套了三層，所以假設有1000根木棍，那就要進行10^9次三角形測試
唉... 真是有毒的迴圈啊!於是我拿起紙筆開始計算

最後，皇天不負苦心人，我導出了下面這條式子
```python
num = min(n,(i+j-1))-j
```
簡單來說，只要知道 i， j，和棒子總數，就可以直接知道有幾個可組成的三角形

一樣再加上一些小小的細節
```python
def count_triangle(n): 
    num = 0 
 
    for i in range (1,n-1): 
        for j in range(i+1,n): 
            num += min(n,(i+j-1))-(j) 
    print (num) 
 
turn = int(input()) 
li = input().split() 
 
for i in range (turn): 
    count_triangle(int(li[i]))
```
正當我一邊陶醉在自己的聰明才智，一邊又從一旁的馬克杯中啜飲著一口白開水
令人心態崩的一幕又來了,依然是有紅的(有點忘了是4綠1紅還是3綠2紅了)

## 第三次解題
不...甚麼哭...誰在哭!
那桶水...那桶水...只是我眼睛流的汗而已
你也知道寫程式碼是...是很傷眼睛的

反正冷靜下來之後，問題出在哪呢?
對! 就是那該死的出題老師，很明顯的出題老師就是來亂的
故意出一個超級大的數字，然後又用垃圾場撿來的電腦跑
(開玩笑的，請不要當真)

但在解決出題老師前，我們還是要解決這一題
很明顯迴圈的毒還是在，從三個到兩個只是減輕毒性而已

於是我伸出沾滿淚...我是說汗水的手，又拿起了紙和筆

有是一番的計算，最後終於算出了下列的關係式
```python
if (n-2*i)>0 : 
            num = (n-2*i)*(i-1)+(i*(i-1))/2 
else: 
            num = (n-i)*(n-i-1)/2
```

簡單來說，這次只要知道 i，和棒子總數，就可以直接知道有幾個可組成的三角形

一樣再加上一些小小的細節

```python
    def count_triangle(n): 
        num = 0 
     
        for i in range (1,n): 
            if (n-2*i)>0 : 
                num += (n-2*i)*(i-1)+(i*(i-1))/2 
            else: 
                num += (n-i)*(n-i-1)/2 
     
        print (int(num)) 
     
    turn = int(input()) 
    li = input().split() 
     
    for i in range (turn): 
        count_triangle(int(li[i]))  
```

正當我一邊處在理智線斷裂的邊緣，一邊又從一旁的馬克杯中啜飲著一口帶鹹味的水
令人心態崩，噢不是令人振奮的一幕來了,終於出現了五綠全對的結果了!

當下我在房間裡笑的像個沒學過Python的孩子
笑的那麼天真，那麼的活潑開朗，沒有一絲壓力

## 結語
以上就是有關我和C_GM01纏鬥一個下午的故事
最後呢還是成功的征服的這個題目
不過就是理智有些模糊，手有些抖，淚腺...不，是汗腺有些腫而已(外加一半的器官衰竭)

而最後的完全體，只要知道棒子總數，就可以直接推得有幾個可組成的三角形
我到現在也不知道，期待大神降臨留言區吧 !

**Comment by evanchou1204** [comment]
故事好精彩！喜歡
阿但是可以解釋一下最上面那個我看得懂的迴圈是怎麼變成下面那個不知道在寫什麼的關係式嗎😢

**Reply by eason8479** [reply]
在打故事的時候省略的推導過程，主要是怕故事太長
還有就是怕看到一堆公式會嚇跑一堆人
不過既然有人問了，下面附上推導(可能有些小複雜)
(此為復刻版，原稿可能已被某資源回收廠銷毀)

第二次解題
![Image](https://i.imgur.com/e5woMnZ.jpg)

第三次解題

建議先看這個
![Image](https://i.imgur.com/Tfr8hwd.jpg)
再看這個
![Image](https://i.imgur.com/LBC0H1U.png)

**Reply by jdlin0652** [reply]
try :
    turn = int(input())
    li= input().split()
    
    for i in range ( turn):
        num = 0
        n = int(li[i])
        for i in range (1,n):
            # num += (n-2*i)*(i-1)+(i*(i-1))/2 if (n-2 * i) > 0 else  (n-i) * (n - i- 1)/2
            num += (n-2*i)*(i-1)+(i*(i-1))/2 if (n > 2*i) else  (n-i) * (n - i- 1)/2
        print (int(num))
except :
    pass

請問貼圖上兩種狀況判斷  n > 2*i  or   n < 2*i 
可你使用的判斷條件 (n-2 * i) > 0 , 是否有特別用意呢?

**Reply by eason8479** [reply]
其實也沒有什麼特別的用意，主要就是依照當下想到的去寫。
兩種寫法是等價的，對程式的運算來說，應該也不會有影響。

**Comment by Chiway1229** [comment]
挖，看到你的心路歷程
讓我想到高中時刷題的痛苦回憶
不過每當他亮起AC時，就會覺得自己又成長了一點了呢

**Comment by Doreen331** [comment]
一樣是程式的奴隸同路人，給你加加油
你真的很有耐心還慢慢推倒，我寫不出來都直接上stackoverflow爬爬爬😃

**Comment by StevenWanglolz** [comment]
這故事好精采哈哈哈~如果是我絕對沒有那個耐心坐一個下午，一定直接上網查解答!

**Comment by BlahFish1006** [comment]
這個故事有聲音哈哈哈好有趣，好佩服你的意志力，感覺已經很久沒有靠自己好好思考一個問題了🥲
但努力一個下午終於靠自己征服一個題目，那感覺真的是格外舒爽啊～
一個下午換來開心一整天 完全值得啊啊啊
